File: controller.py
import time
import os
import psutil

VMRUN = r"Vmrun path give here which u find in your C drive"

VM2 = r"VirtualMachinespath u have to give here which end like -->\app-vm-2\app-vm-2.vmx"
VM3 = r"VirtualMachinespath u have to give here which end like -->\app-vm-3\app-vm-3.vmx"

UPPER = 70
VERY_HIGH = 85
LOWER = 25
VERY_LOW = 15

CHECK_INTERVAL = 5

vm2_running = False
vm3_running = False


def cpu_usage():
    return psutil.cpu_percent(interval=1)


def start_vm(path):
    os.system(f'"{VMRUN}" start "{path}"')

def suspend_vm(path):
    os.system(f'"{VMRUN}" suspend "{path}"')


while True:
    cpu = cpu_usage()
    print(f"CPU: {cpu:.2f}%")

    if cpu > UPPER and not vm2_running:
        print("Starting VM2")
        start_vm(VM2)
        vm2_running = True

    if cpu > VERY_HIGH and not vm3_running:
        print("Starting VM3")
        start_vm(VM3)
        vm3_running = True

    if cpu < LOWER and vm3_running:
        print("Suspending VM3")
        suspend_vm(VM3)
        vm3_running = False

    if cpu < VERY_LOW and vm2_running:
        print("Suspending VM2")
        suspend_vm(VM2)
        vm2_running = False

    time.sleep(CHECK_INTERVAL)

File: controller.java
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import com.sun.management.OperatingSystemMXBean;
import java.lang.management.ManagementFactory;

public class AutoScaleVM {

    // Give your vmrun.exe full path here
    static String VMRUN = "C:\\Path\\To\\vmrun.exe";

    // Give your .vmx file paths here
    static String VM2 = "C:\\VirtualMachines\\app-vm-2\\app-vm-2.vmx";
    static String VM3 = "C:\\VirtualMachines\\app-vm-3\\app-vm-3.vmx";

    static final int UPPER = 70;
    static final int VERY_HIGH = 85;
    static final int LOWER = 25;
    static final int VERY_LOW = 15;

    static final int CHECK_INTERVAL = 5; // seconds

    static boolean vm2Running = false;
    static boolean vm3Running = false;

    public static void main(String[] args) {
        while (true) {
            double cpu = getCpuUsage();
            System.out.printf("CPU: %.2f%%%n", cpu);

            try {
                if (cpu > UPPER && !vm2Running) {
                    System.out.println("Starting VM2");
                    startVM(VM2);
                    vm2Running = true;
                }

                if (cpu > VERY_HIGH && !vm3Running) {
                    System.out.println("Starting VM3");
                    startVM(VM3);
                    vm3Running = true;
                }

                if (cpu < LOWER && vm3Running) {
                    System.out.println("Suspending VM3");
                    suspendVM(VM3);
                    vm3Running = false;
                }

                if (cpu < VERY_LOW && vm2Running) {
                    System.out.println("Suspending VM2");
                    suspendVM(VM2);
                    vm2Running = false;
                }

                TimeUnit.SECONDS.sleep(CHECK_INTERVAL);

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    // Get CPU usage %
    public static double getCpuUsage() {
        OperatingSystemMXBean osBean =
                (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();

        double load = osBean.getSystemCpuLoad();

        if (load < 0) {
            return 0;
        }

        return load * 100;
    }

    // Start VM
    public static void startVM(String path) throws IOException {
        ProcessBuilder pb = new ProcessBuilder(VMRUN, "start", path);
        pb.inheritIO();
        pb.start();
    }

    // Suspend VM
    public static void suspendVM(String path) throws IOException {
        ProcessBuilder pb = new ProcessBuilder(VMRUN, "suspend", path);
        pb.inheritIO();
        pb.start();
    }
}v
